core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1001 @@
+      }
+    }
+  }
+
+  /**
+   * Tracks additional information about a set of equivalent addresses needed for outlier
+   * detection.
+   */
+  static class AddressTracker {
+
+    private final OutlierDetectionLoadBalancerConfig config;
+    private CallCounter activeCallCounter = new CallCounter();
+    private CallCounter inactiveCallCounter = new CallCounter();
+    private Long ejectionTimeNanos;
+    private int ejectionTimeMultiplier;
+    private final Set<OutlierDetectionSubchannel> subchannels = new HashSet<>();
+
+    AddressTracker(OutlierDetectionLoadBalancerConfig config) {
+      this.config = config;
+    }
+
+    boolean addSubchannel(OutlierDetectionSubchannel subchannel) {
+      // Make sure that the subchannel is in the same ejection state as the new tracker it is
+      // associated with.
+      if (subchannelsEjected() && !subchannel.isEjected()) {
+        subchannel.eject();
+      } else if (!subchannelsEjected() && subchannel.isEjected()) {
+        subchannel.uneject();
+      }
+      return subchannels.add(subchannel);
+    }
+
+    boolean removeSubchannel(OutlierDetectionSubchannel subchannel) {
+      return subchannels.remove(subchannel);
+    }
+
+    boolean containsSubchannel(OutlierDetectionSubchannel subchannel) {
+      return subchannels.contains(subchannel);
+    }
+
+    @VisibleForTesting
+    Set<OutlierDetectionSubchannel> getSubchannels() {
+      return ImmutableSet.copyOf(subchannels);
+    }
+
+    void incrementCallCount(boolean success) {
+      // If neither algorithm is configured, no point in incrementing counters.
+      if (config.successRateEjection == null && config.failurePercentageEjection == null) {
+        return;
+      }
+
+      if (success) {
+        activeCallCounter.successCount.getAndIncrement();
+      } else {
+        activeCallCounter.failureCount.getAndIncrement();
+      }
+    }
+
+    @VisibleForTesting
+    long activeVolume() {
+      return activeCallCounter.successCount.get() + activeCallCounter.failureCount.get();
+    }
+
+    long inactiveVolume() {
+      return inactiveCallCounter.successCount.get() + inactiveCallCounter.failureCount.get();
+    }
+
+    double successRate() {
+      return ((double) inactiveCallCounter.successCount.get()) / inactiveVolume();
+    }
+
+    double failureRate() {
+      return ((double)inactiveCallCounter.failureCount.get()) / inactiveVolume();
+    }
+
+    void clearCallCounters() {
+      activeCallCounter.successCount.set(0);
+      activeCallCounter.failureCount.set(0);
+      inactiveCallCounter.successCount.set(0);
+      inactiveCallCounter.failureCount.set(0);
+    }
+
+    void decrementEjectionTimeMultiplier() {
+      // The multiplier should not go negative.
+      ejectionTimeMultiplier = ejectionTimeMultiplier == 0 ? 0 : ejectionTimeMultiplier - 1;
+    }
+
+    void resetEjectionTimeMultiplier() {
+      ejectionTimeMultiplier = 0;
+    }
+
+    void swapCounters() {
```

Comment:
I don't think this function accomplishes what it needs to accomplish. The spec doesn't always fully describe this operation, but in the [call counter section](https://github.com/grpc/proposal/blob/master/A50-xds-outlier-detection.md#call-counter) it says "When the timer triggers, the inactive bucket is zeroed and swapped with the active bucket." The purpose of this operation is to atomically switch the active bucket into inactive mode and make a fresh empty bucket active, making sure not to miss any calls during the transition.

Reply:
Thanks for looking into this code! The zeroing out of the inactive bucket was indeed missing, I added that now.

I did not make the swap atomic, as this method should only be invoked from a timer that is scheduled with a fixed delay. There should never be more than one thread calling this method at the same time.

Reply:
I did make one change to this by making the active counter volatile. This will prevent picker threads from holding on to a stale reference that still points to the inactive one that they would end up incorrectly incrementing. But the swap method is still itself not threadsafe and only called by the timer thread.

Reply:
It's not a synchronization issue, but rather that the inactive counters are not being reset to 0 before they are made active so they are the sum of all previous counts since the last ejection rather than for the current interval.